### ISSUE TITLE
Documentation Request regarding Certificates

### ISSUE THREAD

**Issue by Dalacor** [DCourtel/Wsus_Package_Publisher#91]
I have recently moved to using Certificates created by Openssl for all services that require certificates. The main reason was so that I can use elliptic curves, but also so that I can control the time length of the certificate.

Yesterday, was the first time I needed to update the certificates for WSUS, so decided to do both WSUS and WSUS Package Publisher. I have managed to get everything working, but I did encounter two issues that are not documented in the Wiki Guide. And a third issue that I have not raised yet.

**Issue one:**

To create a certificate for WSUS, I need to install the pfx certificate (certificate and key) in Trusted Root Certification Authorities and in Web Hosting for the WSUS Server. I then need to deploy just the certificate to Trusted Root Certification Authorities for the clients. I need to find out somewhere whether I need the key and certificate for both Web Hosting and Trusted Root Certification Authorities on the WSUS Server.

For WSUS Package Publisher, looking at the documentation, it would seem that I need to import the pfx Certificate into WSUS store, Trusted Root Certification Authorities and Trusted Publishers Store on the Server running WSUS/WSUS Package Publisher. Then you need to deploy just the certificates to the Trusted Root Certification Authorities and Trusted Publishers Store for the clients. As with WSUS, the question is whether you need the key and certificate for WSUS store, Trusted Root Certification Authorities and Trusted Publishers Store on the Server. It's also not clear if you need to import the certificate into WSUS store on the server - I only do that because the picture in the documentation suggests that you need to.

Can you clarify exactly which stores the certificate and key needs to be imported into on the Server running WSUS/WSUS PP. Just to be clear, I created a self signed certificate using Openssl.

**Issue two:**

I discovered after replacing the certificate for both WSUS and WSUS Package Publisher, that I was able to add new updates to WSUS PP, but existing packages that were already deployed, came up with a signing error if you tried to revise it or expire it. I ended up having to decline the updates, delete them and create new updates for these packages all over again.

What is the procedure to updating certificates for WSUS Package Publisher to avoid the need to delete all the existing packages and re-create them? I understand why the error is occurring. I just wondered if there is someway to re-sign the files using the new certificate.

**Issue Three**

I have discovered a few months back, when I retired a client, even though I removed WSUS Package Publisher from their servers, their WSUS Server continued to deploy the software that had been setup using WSUS Package Publisher which meant that the anti virus solution that I provided, was being re-deployed again. The solution was to delete everything in the WSUS Content folder and WSUS UpdateServicesPackages folder and that seemed to prevent the packages from being installed again. I can see that the UpdateServicesPackages is used by WSUS Package Publisher and I am able to clear deleted updates from that folder within WSUS Package Publisher. I know that the WSUS Content folder is used by WSUS itself for all updates from Microsoft and from WSUS Package Publisher.

Interestingly enough, when I was trying to fix the problem with the existing updates not working, I thought the quickest solution would be to delete WSUS Package Publisher and recreate the program. I deleted the files within UpdateServicesPackages as well as the specific folders in the Content folder being used by each package. However, when loading up the new WSUS Package Publisher and connected to the server all the existing packages will still showing - so it's clearly getting information from the WSUS database itself. Unfortunately I couldn't remove the updates from WSUS itself, because they don't show there by default.

So the question here is, what is the procedure to properly remove updates deployed from WSUS Package Publisher from WSUS itself. All the package/cabinet files created over the years are not deleted in WSUS Content folder and clearly is not deleted from the WSUS database.

WSUS Package Publisher needs a Server Cleanup tool like WSUS to remove old packages that are no longer being deployed by WSUS Package Publisher or the delete function needs to be updated to delete the cabinet files from both WSUS folders and from the WSUS database.

**Comment by Dalacor** [comment]
I forgot to ask about Common Name and SAN for WPP Certificates. I assume the SAN should be the FQDN of the WSUS/WSUS Package Publisher Server and Common Name can be anything like WPP for example?

**Comment by Jetsord** [comment]
Maybe that my answer is not 100% correct, but the following workflow works for me.
I have an own (selfsign) Root-CA -> have to be for the WSUS and clientsystems in the "Trusted Root Certification Authorities" store. 
The WSUS Server Certificate (Child of Root- or Sub-CA) is a very standard certificate with "SSL-Webserver"-Flag and FQDN as SAN and (e. g. only) hostname as Common-Name.
The WSUS PP Certificate seems to me, needs nothing special. I give it just the "CodeSigning"-Flag and name the certificate with the generation in the name (e.g. "WSUS PP CodeSigning G1"), import the certificate in the WSUS PP Tool and push the public key into the "Trusted Publishers" Store of every client system. 
So through the import function the CodeSigning certificate is in almost every certificate store. This is not really beautiful, but it is only on the WSUS PP system, so I can tolerate it.
I handle the WSUS PP CodeSigning Certificate like a Root-Certificate and give a new version a other name (-> "G2"), so I do not delete the old one and it can still validate the old updates.

**Comment by Dalacor** [comment]
The difference between WSUS and WSUS Package Publisher Certificate is that WSUS is ServerAuth and WPP is Code Signing in extended key usage. Other than that, they are pretty much identical.

I create one certificate for WSUS which I am placing in Trusted Certification Authorities and Web Hosting. I see what you have done is create a CA that you place in Trusted Certification Authorities for servers and clients and then a certificate for WSUS Server which presumably you have placed in web hosting on the WSUS Server itself.

Because I combined the root CA and Server Certificate into one certificate as it were - the process is not as obvious. So essentially with WSUS you are saying put Certificate into Trusted Root Certification Authorities for Root and Web Hosting for Server Cerrtificate as IIS will not find certificate if it is not in Web Hosting. For the clients, because they are not the WSUS Server, they don't need the Server Certificate which would be found in Web Hosting. It is more clear if you have two certificates - I had not fully appreciated that I was effectively combing the root and server certificates as it were into one certificate.

For WPP, I presume the same process is applied. One Certificate - that is actually two certificates Root and Server Certificate. So needs to be in Trusted Root Certification Authorities for the Root CA. I would imagine for the Server Certificate part of the certificate that it doesn't make any difference if it goes in WSUS or Web Hosting on the Server as it would presumably be the same thing - a Server Certificate. WPP Documentation shows use WSUS - but it should work in Web Hosting as well?

I am unclear as to why the Certificate needs to go into Trusted Publishers on both Servers and Clients - as I would assume if it was in the Trusted Root Certification Authorities - this effectively is the same thing - so having it in Trusted Publishers would be redundant? Trusted Publishers Is certainly doesn't needed for WSUS itself.

I think would need the keys and certificates for both Certificates on the Server - but only the certificates on the clients.

So I will amend my question - why is WSUS only needing CA Certificate in root certification authorities on all machines, but WPP seems to require the CA Certificate in both Root Certification Authorities and Trusted Publishers - it should be one or the other.

I like the idea of naming WPP something like WPP 2024 or WPP G1 in the SAN. That makes sense.

I will have to check next time and see whether I deleted the old certificates from the Server itself or whether I just updated the Certificate within WSUS Package Publisher. I think what I will do next time is simply delete the packages, then update the certificates and recreate the packages because I only have a few packages and it means I start with a recent versions of deployments. It will only take me 10 minutes. However, I will use generational certificates as you have suggested.

**Comment by Jetsord** [comment]
I can only base my guess on how I find it in the Windows certificate store. The Trusted Publishers store always includes the specific codesigning certificates and not just the root certificate from which it originates. But the root certificate have to be in the Trusted Root Authorities as well. I also believe that the Windows certificate store folders are more of a logical sorting and not a major technical differentiation. So I added initial the WPP certificate to the Publishers Store.

The WSUS only needs the WPP certificate in the Trusted Publishers Store if a WSUS-WPP package is to be installed. WSUS PP "packs" the setup into a signed cab-file and is only checked during installation. Or like my first assumption, the WPP codesigning certificate can be in any certificate store.

And the WSUS itself is a "pure" web server and there logic is usually the same: if I trust the root CA from which the web server comes, I trust the site. (In this case it is just for the secure file transfer between WSUS and client system)

For signed installation files, the certificate-trust-relationship is apparently stricter. 

And the downloaded and already signed cab-files of the Microsoft/Windows updates are not re-signed by the local WSUS, so some preinstalled Microsoft certificates from the Trusted Publisher Store validates the installations.

**Comment by Jetsord** [comment]
So I recommend a root-CA infrastructure for small environments the tool "XCA", 
this make the certificate structure a bit easier. So the structure is this (to clearify and help somebody else): 

**XCA-Tool** (can be saved "everywhere")
"My Root CA" with CA-flag
_child-certificates:_
"WSUS-FQDN" with Serverauthentication-flag 
"WSUS PP CodeSigning G1" with Codesigning-flag


**WSUS Server**
Trusted Root Authories | "My Root CA"
Personal Certificate | "WSUS-FQDN" (with private key) -> _linked in IIS_
WSUS PP | "WSUS PP CodeSigning G1" (with private key) -> _in many stores through the WSUS PP Import function_

**Client System**
Trusted Root Authories | "My Root CA"
Trusted Publishers | "WSUS PP CodeSigning G1"

**Comment by Dalacor** [comment]
Thank you. That is very concise. I have updated my documentation to reflect pretty much the above - although my wsus certificate goes in web hosting, not personal - but it doesn't really make much difference. It doesn't harm anything to have the WSUS WPP certificate in Trusted Publishers as well, but interesting that it is required when WSUS itself does not. But I have updated the documentation accordingly.

Thank you.
